rfcs/text/0000-prefab-structs.md

```diff
@@ -0,0 +1,336 @@
+producing the following type definition:
+
+`(define-type point (Prefab point Integer Integer))`
+
+and the following functions:
+
+- `point` of type `(-> Integer Integer point)`
+
+- `point?` of type `(-> Any Boolean : point)` _NOTE: this is unsound since
+  the fields could contain anything_
+
+- `point-x` of type `(-> point Integer)` (this is sound but too limited, i.e.
+it won't work for `point`'s with different field types (and it should))
+
+- `point-y` of type `(-> point Integer)` (same issue as `point-x`)
+
+### Racket v6.12 Mutable Prefabs
+
+```racket
+(struct initials ([first : Symbol] [last : Symbol])
+  #:prefab
+  #:mutable)
+```
+
+defines types/constructors/accessors for a mutable prefab struct, 
+producing the following type definition:
+
+`(define-type initials (Prefab (initials #(0 1)) Symbol Symbol))` 
+
+and the following function definitions:
+
+- `initials` with type `(-> Symbols Symbols initials)`
+
+- `initials?` with type `(-> Any Boolean : initials)` (_NOTE: this is unsound
+since the fields can actually have values of any type in them_)
+
+- `initials-first` with type `(-> initials Symbol)` (_NOTE: this is likely sound
+but too limited like the current immutable field accessors_)
+
+- `initials-last` with type `(-> initials Symbol)` (_NOTE: same as previous accessor_)
+
+- `set-initials-first!` with type `(-> initials Symbol Void)` (_NOTE: also too limited_)
+
+- `set-initials-last!` with type `(-> initials Symbol Void)` (_NOTE: also too limited_)
+
+## Backwards Compatibility
+
+- Any previous programs that relied on a prefab's predicate and occurrence typing 
+to determine not only that a value was a prefab, but also that it's fields
+had values of a certain type, will break. 
+
+E.g. this program will break:
+
+```racket
+(ann (lambda (p) (if (point? p) (point-x p) 42))
+     (-> Any Integer))
+```
+
+I.e. we know `p` is a `point` _prefab_ of some sort in the then branch,
+but not that it's fields have integers in them. 
+
+This program will still work:
+
+```racket
+(ann (lambda (p) (if (point? p) (point-x p) p))
+     (-> (U point Integer) Integer))
+```
+
+because we knew that `p` had type `point` OR it had type `Integer`, 
+so the test for `point?` producing `#true` indicates it was not 
+an `Integer` and is indeed a `point` _prefab_ with Integers in both its fields.
+
+- Changing the accessors to be polymorphic could break programs that work
+today but where our current limited type inference would fail (e.g. mapping
+a struct accessor over a list of structs). We could mitigate this
+by using a `case->` type with the old type after the polymorphic type if
+we think it's worth it.
+
+E.g. today this program type checks:
+
+```racket
+(map point-x
+     (list (point 1 2) (point 3 4)))
+```
+
+but will break if we only use the polymorphic type for accessors 
+(since our limited inference cannot infer types for polymorphic 
+arguments to polymorphic functions)
+
+# Drawbacks and Alternatives
```

Comment:
This section should discuss whether Typed Racket could change the behavior of `struct` to produce functions that work with the types we've been giving them. 

I think the answer is: yes for immutable prefabs whose fields are all first-order data, no in other cases.